### PR TITLE
Conv+Activation fusion for CPU

### PR DESCRIPTION
**Pull request by duli2012**
OP fusion for conv+activations, currently support LeakyRelu, Relu, Sigmoid, Tanh.

### CODE REVIEW

**Comment by weixingzhang** [comment]
Did you try any model and see performance improvement?

**Comment by duli2012** [comment]
@weixingzhang Tracy tried is out on a model we are working on and we did see performance improvement.

File: onnxruntime/core/graph/conv_activation_fusion.cc [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+}  // namespace
+
+Status ConvActivationFusion::Apply(Graph& graph, bool& modified) const {
+  GraphViewer graph_viewer(graph);
+  const auto& order = graph_viewer.GetNodesInTopologicalOrder();
+
+  std::vector<onnxruntime::NodeIndex> removed_nodes;
+  for (auto index : order) {
+    auto node = graph.GetNode(index);
+    if (node->OpType() != "Conv" || node->GetOutputEdgesCount() != 1) {
```
Comment (tracysh) [inline onnxruntime/core/graph/conv_activation_fusion.cc]: This should call the new utils::IsSupportedOptypeVersionAndDomain. Same for IsFusableActivation

File: onnxruntime/core/graph/conv_activation_fusion.cc [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    const Node& next_node = *(node->OutputNodesBegin());
+    if (!IsFusableActivation(next_node) || graph.IsNodeOutputsInGraphOutputs(next_node)) {
+      continue;
+    }
+
+    Node* conv_node = node;
+    const Node& act_node = next_node;
+    std::vector<NodeArg> input_args, output_args;
+
+    Node& fused_conv = graph.AddNode("fused " + conv_node->Name(), "FusedConv",
```
Comment (tracysh) [inline onnxruntime/core/graph/conv_activation_fusion.cc]: DML has code to generate unique names too with a comment about the node name (conv_node in this case) being optional. Does more need to be done here to guarantee uniqueness? Is there some algorithm that maybe DML and these fusions could share?

File: onnxruntime/core/graph/conv_activation_fusion.cc [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                                     conv_node->MutableInputDefs(),
+                                     conv_node->MutableOutputDefs(),
+                                     &conv_node->GetAttributes(),
+                                     "com.microsoft");
+
+    //Add a new attribute to specify the activation type
+    AttributeProto act;
+    act.set_name("activation");
+    act.set_type(ONNX_NAMESPACE::AttributeProto_AttributeType_STRING);
+    act.set_s(act_node.OpType());
```
Comment (tracysh) [inline onnxruntime/core/graph/conv_activation_fusion.cc]: DML uses Node->AddAttribute("activation", "string") which internally does all this stuff.

File: onnxruntime/contrib_ops/cpu/fused_conv.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+#include "core/providers/cpu/nn/conv_impl.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+template <typename T>
+class FusedConv : public Conv<T> {
+ public:
+  FusedConv(const OpKernelInfo& info) : Conv<T>(info) {
+    std::string activation_type;
```
Comment (tracysh) [inline onnxruntime/contrib_ops/cpu/fused_conv.h]: activation_type is not used here?

File: onnxruntime/core/providers/cpu/nn/conv.cc [hunk 5]
```diff
@@ -4,9 +4,153 @@
...
+      if (B != nullptr) {
+        auto Ymatrix = EigenMatrixMap<float>(Ydata, output_image_size, M);
+        auto Bvec = ConstEigenVectorMap<float>(B->template Data<float>(), M);
+        Ymatrix.rowwise() += Bvec.transpose();
+      }
+
+      Xdata += X_offset * group_;
+      Ydata += Y_offset * group_;
+    }
+  }
```
Comment (tracysh) [inline onnxruntime/core/providers/cpu/nn/conv.cc]: fuse_activation needs to be called for both paths here (2D/3D and the other cases). I'll remove the fuse_activation from the MLAS path when I check in my changes to use it.

File: onnxruntime/core/graph/conv_activation_fusion.cc (hunk 1 above)
Comment (duli2012) [reply onnxruntime/core/graph/conv_activation_fusion.cc]: Done

File: onnxruntime/core/graph/conv_activation_fusion.cc (hunk 3 above)
Comment (duli2012) [reply onnxruntime/core/graph/conv_activation_fusion.cc]: Replaced.

File: onnxruntime/contrib_ops/cpu/fused_conv.h (hunk 4 above)
Comment (duli2012) [reply onnxruntime/contrib_ops/cpu/fused_conv.h]: typo. should be "activation"

File: onnxruntime/core/providers/cpu/nn/conv.cc (hunk 5 above)
Comment (duli2012) [reply onnxruntime/core/providers/cpu/nn/conv.cc]: Added.

File: onnxruntime/core/graph/conv_activation_fusion.cc (hunk 2 above)
Comment (duli2012) [reply onnxruntime/core/graph/conv_activation_fusion.cc]: Fixed.

File: onnxruntime/contrib_ops/cpu/fused_conv.h (hunk 4 above)
Comment (duli2012) [reply onnxruntime/contrib_ops/cpu/fused_conv.h]: Ah. Now both fixed.

**Pull request merged**